### DISCUSSION TITLE
Fill-mask language model

### DISCUSSION THREAD

**Discussion by agistaputri20** [deepjavalibrary/djl#2421] [Q&A] [answered]
Does anyone has an example on how to run inference a fill-mask language modeling on Scala/Java using DJL?

**Comment by frankfliu** [comment]
We don't have an end2end example, but you can find sample code in our unit test: https://github.com/deepjavalibrary/djl/blob/master/extensions/tokenizers/src/test/java/ai/djl/huggingface/tokenizers/FillMaskTranslatorTest.java#L90-L108

It's very simple if you use DJL model zoo:

```
        Criteria<String, Classifications> criteria =
                Criteria.builder()
                        .setTypes(String.class, Classifications.class)
                        .optModelUrls("djl://ai.djl.huggingface.pytorch/bert-base-uncased")
                        .optEngine("PyTorch")
                        .build();

```

**Reply by agistaputri20** [reply]
Alright thank you! But how to make that ModelUrl from huggingface model? Because I have my own model on huggingface hub.

**Reply by agistaputri20** [reply]
And when I use my own model, do I have to make a translator that contain a preprocess and postprocess function?

**Comment by frankfliu** [comment]
@agistaputri20 

1. For Huggingface fill-mask model, we have built-in translators that you can use, they handle the pre-processing and porst-processing for you. You don't need create your own translator. You have want to some special handling, you can use our [FillMaskTranslator](https://github.com/deepjavalibrary/djl/blob/master/extensions/tokenizers/src/main/java/ai/djl/huggingface/translator/FillMaskTranslator.java) as base and modify on top of it.
2. DJL support batch inference, but you need set padding/truncation for tokenizer.
3. You use this script to import your hugingface model: https://github.com/deepjavalibrary/djl/blob/master/extensions/tokenizers/src/main/python/model_zoo_importer.py, the output .zip file is ready to use in DJL, you only need specify TranslatorFactory: 

```
        Criteria<String, Classifications> criteria =
                Criteria.builder()
                        .setTypes(String.class, Classifications.class)
                        .optModelPath(Paths.get("YOUR_ZIP_FILE.zip"))
                        .optEngine("PyTorch")
                        .optTranslatorFactory(new FillMaskTranslatorFactory())
                        .build();
```

**Reply by agistaputri20** [reply]
<img width="598" alt="image" src="https://user-images.githubusercontent.com/100820635/222647338-f46ae017-86ae-43c6-b4a3-b25c2d444316.png">
 When I try to use your translator and convert it to Scala, I got errors like this, could you explain to me how to solve it?

**Reply by frankfliu** [reply]
Why you need convert it? Why not just directly use it as is?

should it be:

```
maskTokenId = encoding.getIds()[0]
```

**Reply by agistaputri20** [reply]
Because I have a task to run inference a fill-mask model on Scala using DJL, Sir.

<img width="610" alt="image" src="https://user-images.githubusercontent.com/100820635/222751942-3cbc87a5-6ef9-4a7a-bde5-8855e8f01bfa.png">
I have change it but still doesn't work

**Reply by frankfliu** [reply]
Scala can call Java class directly.
The error is because you declared maskTokenId as Int, but the return values is long

**Reply by agistaputri20** [reply]
So when I use Scala to run inference, I don't have to convert all the structure on that translator? Just by calling them on criteria builder, is it correct?

**Reply by frankfliu** [reply]
Yes, you can use Criteria and load the model directly in Scala. What you need is just include huggingface tokenizer in your project:

https://docs.djl.ai/master/extensions/tokenizers/index.html#installation

**Reply by agistaputri20** [reply]
<img width="521" alt="image" src="https://user-images.githubusercontent.com/100820635/223948198-ab4d66ee-e118-496d-95c6-aa4a5610b904.png">
 In this section should I download the master DJL folder on my local first, so I can get the path then try to run it?

**Reply by frankfliu** [reply]
Yes, that's the easiest way to convert a huggingface FillMask model to DJL.

**Reply by agistaputri20** [reply]
Great! I already have my huggingface zip model

Then, I've tried to run my code but why this is happen? Even though I have imported the ai.djl.huggingface.translator.FillMaskTranslator and huggingface dependency on my build.sbt file. So, what should I add on my program?
<img width="774" alt="image" src="https://user-images.githubusercontent.com/100820635/224223044-fb05bf2e-959b-4adf-8fa2-83a210d7c1a5.png">

**Reply by frankfliu** [reply]
You need remove `.optTranslator(new FillMaskTranslator())` in your code and replace with `.optTranslatorFactory(new FillMaskTranslatorFactory())`

FillMaskTranslator's constructor is not public, you can not use `new FillMaskTranslator()`, you have to use FillMaskTranslator.builter() ... to create a new FillMaskTranslator instance. Or use preferred FillMaskTranslatorFactory way.

**Comment by agistaputri20** [comment]
I have to use this model from huggingface: https://huggingface.co/jibi2906/my-finetuned-distilbert-pt. But it makes error like this, do you know how to solve it?
<img width="908" alt="image" src="https://user-images.githubusercontent.com/100820635/224243972-7ff2fb6a-a09c-4827-80b4-dd2b7b34de86.png">

**Reply by frankfliu** [reply]
I updated my earlier reply. You need use `.optTranslatorFactory(new FillMaskTranslatorFactory())`

**Reply by agistaputri20** [reply]
Nice, it works! Thank you :)

**Reply by agistaputri20** [reply]
Wait, sorry is it possible to detect and fill more than one [MASK]?

**Reply by frankfliu** [reply, marked as answer]
No, that's the model's limitation.
